tests/wlr_foreign_toplevel_management_v1.cpp

```diff
@@ -0,0 +1,740 @@
+    wlcs::Client client;
+    wlcs::ForeignToplevelManager manager;
+    wlcs::Surface surface;
+    wlcs::XdgSurfaceStable xdg_surface;
+    wlcs::XdgToplevelStable xdg_toplevel;
+};
+}
+
+TEST_F(ForeignToplevelManagerTest, supports_zwlr_foreign_toplevel_manager_v1)
+{
+    wlcs::Client client{the_server()};
+    wlcs::ForeignToplevelManager manager{client};
+    client.roundtrip();
+}
+
+TEST_F(ForeignToplevelManagerTest, does_not_detect_toplevels_when_test_creates_none)
+{
+    wlcs::Client client{the_server()};
+    wlcs::ForeignToplevelManager manager{client};
+    client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(0u));
+}
+
+TEST_F(ForeignToplevelManagerTest, detects_toplevel_from_same_client)
+{
+    wlcs::Client client{the_server()};
+
+    auto surface{client.create_visible_surface(w, h)};
+
+    wlcs::ForeignToplevelManager manager{client};
+    client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(1u));
+}
+
+TEST_F(ForeignToplevelManagerTest, detects_toplevel_from_different_client)
+{
+    wlcs::Client foreign_client{the_server()};
+    wlcs::Client observer_client{the_server()};
+
+    auto surface{foreign_client.create_visible_surface(w, h)};
+
+    wlcs::ForeignToplevelManager manager{observer_client};
+    observer_client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(1u));
+}
+
+TEST_F(ForeignToplevelManagerTest, detects_toplevel_created_after_manager)
+{
+    wlcs::Client client{the_server()};
+
+    wlcs::ForeignToplevelManager manager{client};
+    client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(0u));
+
+    auto surface{client.create_visible_surface(w, h)};
+
+    client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(1u));
+}
+
+TEST_F(ForeignToplevelManagerTest, detects_multiple_toplevels_from_multiple_clients)
+{
+    wlcs::Client foreign_client{the_server()};
+    wlcs::Client observer_client{the_server()};
+
+    auto foreign_surface{foreign_client.create_visible_surface(w, h)};
+    auto observer_surface{observer_client.create_visible_surface(w, h)};
+
+    wlcs::ForeignToplevelManager manager{observer_client};
+    observer_client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(2u));
+}
+
+TEST_F(ForeignToplevelManagerTest, detects_toplevel_closed)
+{
+    wlcs::Client client{the_server()};
+
+    wlcs::ForeignToplevelManager manager{client};
+    client.roundtrip();
+    EXPECT_THAT(manager.toplevels().size(), Eq(0u));
+
+    {
+        wlcs::Surface other{client};
+        wlcs::XdgSurfaceStable other_xdg{client, other};
+        wlcs::XdgToplevelStable other_toplevel{other_xdg};
+        other.attach_visible_buffer(w, h);
+        client.roundtrip();
+
+        EXPECT_THAT(manager.toplevels().size(), Eq(1u));
+    }
+
+    client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(0u));
+}
+
+TEST_F(ForeignToplevelHandleTest, gets_title)
```

Comment:
Then these ones that are testing `zwlr_foreign_toplevel_handle_v1` can probably use the `ForeignToplevelManager` helper, but should be making direct calls to `zwlr_foreign_toplevel_handle_v1` (yes, even the listener, although we can perhaps make something easier for that? Like have a default stub listener, and then override individual callbacks in there?)